### PR TITLE
[Collections] Add and wire up models for signed package collections

### PR DESCRIPTION
**Pull request by yim-lee**
- Introduce new `SignedCollection` as
  [proposed](https://forums.swift.org/t/package-collection-signing/43855).
- Update logic in `JSONPackageCollectionProvider` to check for signature
- Adjust tests

### CODE REVIEW

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Comment by yim-lee** [comment]
@swift-ci please smoke test

File: Sources/PackageCollectionsModel/PackageCollectionModel+v1.swift [hunk 1]
```diff
@@ -339,3 +338,145 @@ extension PackageCollectionModel.V1.ProductType: Codable {
...
+        }
+        decoder.userInfo[keyPathUserInfoKey] = \PackageCollectionModel.V1.SignedCollection.collection
+        return try decoder.decode(KeyPathValue<PackageCollectionModel.V1.Collection>.self, from: data).value
+    }
+
+    public static func signature(from data: Data, using decoder: JSONDecoder) throws -> PackageCollectionModel.V1.Signature {
+        guard let keyPathUserInfoKey = CodingUserInfoKey(rawValue: Self.keyPathKey) else {
+            throw KeyPathValueError.invalidUserInfo
+        }
+        decoder.userInfo[keyPathUserInfoKey] = \PackageCollectionModel.V1.SignedCollection.signature
```
Comment (abertelrud) [inline Sources/PackageCollectionsModel/PackageCollectionModel+v1.swift:451]: Is the use of a dictionary entry here to support different versions of key signature with the same decoder/encoder?

File: Sources/PackageCollections/Model/Collection.swift [hunk 2]
```diff
@@ -66,8 +71,8 @@ extension PackageCollectionsModel {
             packages: [Package],
             createdAt: Date,
             createdBy: Author?,
-            lastProcessedAt: Date = Date(),
-            isSigned: Bool
+            signature: SignatureData?,
+            lastProcessedAt: Date = Date()
```
Comment (abertelrud) [inline Sources/PackageCollections/Model/Collection.swift:75]: Does this need a compatibility wrapper callback or are there no clients using this yet?

**Review by abertelrud — APPROVED** [review]

File: Sources/PackageCollections/Model/Collection.swift (hunk 2 above)
Comment (yim-lee) [reply Sources/PackageCollections/Model/Collection.swift:75]: The initializer is not exposed outside of this module so we should be ok.

File: Sources/PackageCollectionsModel/PackageCollectionModel+v1.swift (hunk 1 above)
Comment (yim-lee) [reply Sources/PackageCollectionsModel/PackageCollectionModel+v1.swift:451]: This is for us to decode a specific key (either collection or signature in this case), e.g. we would want to validate the signature before decoding the rest of the bytes. It's looking more like we won't need this code though (i.e., we'd just decode the whole thing), and I will delete it if that's the case.

**Comment by yim-lee** [comment]
@swift-ci please smoke test

**Pull request merged**